kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+		errno = EACCES;
+		return (mqd_t)mqd;
+	}
+
+	/* Allocate message queue descriptor space */
+	if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_desc,
+			     sizeof(struct mqueue_desc), 100) == 0) {
+		memset(block_memq_desc.data, 0, sizeof(struct mqueue_desc));
+		msg_queue_desc = block_memq_desc.data;
+		memcpy(&msg_queue_desc->block, &block_memq_desc,
+		       sizeof(struct k_mem_block));
+	} else {
+		errno = ENOSPC;
+		return (mqd_t)mqd;
+	}
+
+	/* Allocate mqueue object for new message queue */
+	if (msg_queue == NULL) {
+
+		/* Check for message quantity and size in message queue */
+		if (attrs->mq_msgsize > CONFIG_MSG_SIZE_MAX &&
+		    attrs->mq_maxmsg > CONFIG_MSG_COUNT_MAX) {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for message queue object */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_obj,
+				     sizeof(struct mqueue_object), 100) == 0) {
+			memset(block_memq_obj.data, 0,
+			       sizeof(struct mqueue_object));
+			msg_queue = (struct mqueue_object *)block_memq_obj.data;
+			memcpy(&msg_queue->block_mqueue, &block_memq_obj,
+			       sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for block tracking */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_track,
+				     sizeof(struct mem_block),
+				     K_FOREVER) == 0) {
+			memset(block_track.data, 0, sizeof(struct mem_block));
+			msg_queue->block_buffer_name = block_track.data;
+		} else {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for message queue name */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_name,
+				     strlen(name) + 1, K_FOREVER) == 0) {
+			memset(block_memq_name.data, 0, strlen(name)+1);
+			msg_queue->name = block_memq_name.data;
+			memcpy(&msg_queue->block_buffer_name->block_name,
+			       &block_memq_name, sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		strcpy(msg_queue->name, name);
+
+		msg_size = attrs->mq_msgsize;
+		max_msgs = attrs->mq_maxmsg;
+
+		/* Allocate space for message queue buffer */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_buffer,
+				     msg_size * max_msgs * sizeof(u8_t), 100)
+				     == 0) {
+			memset(block_buffer.data, 0,
+			       msg_size * max_msgs * sizeof(u8_t));
+			buffer = block_buffer.data;
+			memcpy(&msg_queue->block_buffer_name->block_buffer,
+			       &block_buffer, sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
+		}
+
+		msg_queue->ref_count = 1;
+		/* initialize zephyr message queue */
+		k_msgq_init(&msg_queue->queue, buffer, msg_size, max_msgs);
+
+		/* Add the message queue at start of list */
+		k_sem_take(&mq_sem, K_FOREVER);
+		msg_queue->prev = NULL;
+
+		if (mq_list != NULL) {
+			mq_list->prev = msg_queue;
+		}
+
+		msg_queue->next = mq_list;
+		mq_list = msg_queue;
+	} else {
+		k_sem_take(&mq_sem, K_FOREVER);
+
+		if (msg_queue->ref_count >= CONFIG_MAX_DESC_PERMQ) {
```

Comment:
Why is there a kconfig for maximum reference count?  What value does that provide?  Shouldn't we allow as many handles to the object as code wants?

Also: this critical section is just a value increment.  Do it with an atomic_inc() call, don't take an expensive semaphore lock and potentially block.

Reply:
For every mq_open() we need to create a mq_descriptor() and we are incrementing ref_count on each descriptor and opening a descriptor needs 16 bytes of space. I had put this config to put an restriction on maximum number of mq_open for a message_queue.  This will help in accounting in memory requirement for message queue too.

ok I will try to make ref_count atomic and avoid semaphore usage.